File: tic_tac_kt/app/bin/main/tic_tac_kt/Board.kt
package tic_tac_kt

inline fun <reified T>Array<Array<T>>.transpose(): Array<Array<T>> {
    return Array(this[0].size) { i -> Array(this.size) { j -> this[j][i] } }
}

class Board {
    companion object {
        const val SIZE = 3
    }

    private var cells: Array<Array<String>> = Array(3) { Array(3) { "" } }

    public fun isValidPosition(row: Int, col: Int, fullCheck: Boolean = true): Boolean {
        if (row < 0 || col < 0 || row >= SIZE || col >= SIZE) {
            return false
        }

        if (fullCheck) {
            return cells.get(row).get(col).isEmpty()
        }

        return true
    }

    public fun getCell(row: Int, col: Int): String {
        if (!isValidPosition(row, col)) {
            throw Exception("Invalid row or column value. Try again")
        }

        return cells.get(row).get(col)
    }

    public fun setCell(row: Int, col: Int, mark: String) {
        if (!isValidPosition(row, col)) {
            throw Exception("Invalid row or column value. Try again")
        }

        cells[row][col] = mark
    }

    public fun display() {
        println("\n     c1  c2  c3")
        println("   -------------")

        cells.forEachIndexed { i, row ->
            print("r${i+1} |")
            row.forEach { cell ->
                print(String.format(" %1s |", cell))
            }

            println("\n   -------------")
        }

        println("")
    }

    public fun isFull(): Boolean {
        return !cells.any { row -> row.any { cell -> cell.isEmpty() } }
    }

    public fun hasHorizontalWin(): Boolean {
        return cells.any { row -> row.distinct().size == 1 && !row.getOrElse(0) { "" }.isEmpty() }
    }

    public fun hasVerticalWin(): Boolean {
        return cells.transpose().any { col -> col.distinct().size == 1 && !col.getOrElse(0) { "" }.isEmpty() }
    }

    public fun hasDiagonalWin(): Boolean {
        val mainDiag = List(SIZE) { i -> cells[i][i] }
        val secDiag = List(SIZE) { i -> cells[i][SIZE - i - 1] }

        val mainMatch = mainDiag.distinct().size == 1 && !mainDiag.getOrElse(0) { "" }.isEmpty()
        val secMatch = secDiag.distinct().size == 1 && !secDiag.getOrElse(0) { "" }.isEmpty()

        return mainMatch || secMatch
    }

    public fun checkWinner(): Boolean {
        return hasHorizontalWin() || hasVerticalWin() || hasDiagonalWin()
    }
    
    public fun setRemainingCells() {
        cells.forEachIndexed { i, row ->
            row.forEachIndexed { j, cell ->
                if (cell.isEmpty()) {
                    setCell(i, j, Utils.Mark.DEF)
                }
            }
        }
    }
}

File: tic_tac_jl/board.jl
module BoardMod

include("utils.jl")

const SIZE = 3

struct Board
    cells::Matrix{String}

    function Board()
        new(Matrix{String}(fill("", SIZE, SIZE)))
    end
end

function is_valid_position(self::Board, row::Int, col::Int, full_check = true)
    if row < 0 || col < 0 row > SIZE || col > SIZE
        return false
    end

    return !full_check || isempty(self.cells[row, col])
end

function get_cell(self::Board, row::Int, col::Int)
    if !is_valid_position(self, row, col, false)
        throw(ErrorException("Invalid row or column value. Try again"))
    end

    return self.cells[row, col]
end

function set_cell(self::Board, row::Int, col::Int, mark::String)
    if !is_valid_position(self, row, col)
        throw(ErrorException("Invalid row or column value. Try again"))
    end

    self.cells[row, col] = mark
end

function display(self::Board)
    println("\n     c1  c2  c3")
    println("   -------------")

    for i in 1:SIZE
        print("r$i |")
        for j in 1:SIZE
            print("$(lpad(self.cells[i, j], 2)) |")
        end

        println("\n   -------------")
    end

    println("")
end

function is_full(self::Board)
    for row in eachrow(self.cells)
        if any(cell -> isempty(cell), row)
            return false
        end
    end
    return true
end

function has_horizontal_win(self::Board)
    for row in eachrow(self.cells)
        if length(Set(row)) == 1 && !isempty(row[1])
            return true
        end
    end
    
    return false
end

function has_vertical_win(self::Board)
    for col in eachcol(self.cells)
        if length(Set(col)) == 1 && !isempty(col[1])
            return true
        end
    end
    
    return false
end

function has_diagonal_win(self::Board)
    main_diag = [self.cells[i, i] for i in 1:SIZE]
    sec_diag = [self.cells[i, SIZE - i + 1] for i in 1:SIZE]

    main_match::Bool = length(Set(main_diag)) == 1 && !isempty(main_diag[1])
    sec_match::Bool = length(Set(sec_diag)) == 1 && !isempty(sec_diag[1])

    return main_match || sec_match
end

function check_winner(self::Board)
    return has_horizontal_win(self) || has_vertical_win(self) || has_diagonal_win(self)
end

function set_remaining_cells(self::Board)
    for i in 1:SIZE
        for j in 1:SIZE
            if isempty(self.cells[i, j])
                set_cell(self, i, j, UtilsMod.DEF)
            end
        end
    end
end

end